### PR TITLE
Editorial rework of streams sections

### PR DESCRIPTION
**Pull request by janaiyengar**
I've tried to eliminate redundancies and rearrange a few things in a better order. This might be hard to read the diffs on, so maybe just look at the Stream sections in the compiled draft.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
+an arbitrary amount of data to be sent on any stream, subject to flow control
+constraints (see {{flow-control}}).
 
-An alternative view of QUIC streams is as an elastic "message" abstraction,
-similar to the way ephemeral streams are used in SST
-{{?SST=DOI.10.1145/1282427.1282421}}, which may be a more appealing description
-for some applications.
+## Stream Types and Identifiers {#stream-id}
 
+<!-- Stream ID or stream ID -->
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: the latter.  We're using capitalization for field names, not concepts.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
+Streams in QUIC provide a lightweight, ordered byte-stream abstraction to an
+application. An alternative view of QUIC streams is as an elastic "message"
+abstraction, similar to the way ephemeral streams are used in SST
+{{?SST=DOI.10.1145/1282427.1282421}}, which may be a more appealing description
+for some applications.
 
-Either type of stream can be created by either endpoint, can concurrently send
-data interleaved with other streams, and can be cancelled.
+QUIC does not provide any means of ensuring ordering between octets on different
+streams.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Too early for this concept.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-Implementations MAY choose to offer the ability to deliver data out of order.
-There is no means of ensuring ordering between bytes on different streams.
-
-Streams are individually flow controlled, allowing an endpoint to limit memory
-commitment and to apply back pressure.  The creation of streams is also flow
-controlled, with each peer declaring the maximum stream ID it is willing to
-accept at a given time.
+QUIC allows for an arbitrary number of streams to operate concurrently and for
+an arbitrary amount of data to be sent on any stream, subject to flow control
+constraints (see {{flow-control}}).
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Maybe the inter-stream ordering belongs here.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
+## Stream Types and Identifiers {#stream-id}
 
+<!-- Stream ID or stream ID -->
 
-## Stream Identifiers {#stream-id}
+Streams can be unidirectional or bidirectional.  Unidirectional streams carry
+data in one direction: from the initiator of the stream to its peer.
+Bidirectional streams allow for data to be sent in both directions.  Streams can
+be created by either endpoint, can concurrently send data interleaved with other
+streams, and can be cancelled.  Any stream can be initiated by either enndpoint.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: This might be better before the header.  Concentrating this bit on identifiers might be easier.

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-STREAM data in frames determined to be lost SHOULD be retransmitted before
-sending new data, unless application priorities indicate otherwise.
-Retransmitting lost stream data can fill in gaps, which allows the peer to
-consume already received data and free up the flow control window.
+resources to, the implementation SHOULD use the information provided by the
+application.  Failure to account for priority of streams can result in
+suboptimal application performance.
 
 
 # Stream States: Life of a Stream {#stream-states}
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
# Stream States {#stream-states}
```

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
+be created by either endpoint, can concurrently send data interleaved with other
+streams, and can be cancelled.  Any stream can be initiated by either enndpoint.
 
-Streams are identified by an unsigned 62-bit integer, referred to as the Stream
-ID.  Stream IDs are encoded as a variable-length integer (see
-{{integer-encoding}}).  The least significant two bits of the Stream ID are used
-to identify the type of stream (unidirectional or bidirectional) and the
-initiator of the stream.
+Streams are identified within a connection by a numeric value, referred to as
+the Stream ID.  Stream IDs are unique to a stream: a QUIC endpoint MUST NOT
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
the Stream ID.  Stream IDs are unique to a stream. A QUIC endpoint MUST NOT
```

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-
-Endpoints MUST NOT exceed the limit set by their peer.  An endpoint that
-receives a STREAM frame with an ID greater than the limit it has sent MUST treat
-this as a stream error of type STREAM_LIMIT_ERROR ({{error-handling}}).
-
-A receiver cannot renege on an advertisement; that is, once a receiver
-advertises a stream limit using the MAX_STREAMS frame, advertising a smaller
-limit has no effect.  A receiver MUST ignore any MAX_STREAMS frame that does
-not increase the stream limit.
+Within each type, streams are expected to be created with numerically increasing
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
Within each type, streams are created with numerically increasing
```

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -420,8 +361,9 @@ conditions for opening a stream are slightly more complex for a bidirectional
 stream because the opening of either send or receive sides causes the stream
 to open in both directions.
 
-An endpoint can open streams up to its maximum stream limit in any order,
-however endpoints SHOULD open the send side of streams for each type in order.
+An endpoint can open streams up to its maximum stream limit in any order of
+stream ID. However, within each type, endpoints SHOULD open the send side of
+streams in numerically increasing order of stream ID.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I think that this text was rotten to begin with.  There might be a simpler statement.

"An endpoint MUST open streams of the same type in increasing order of stream ID."

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -567,37 +512,36 @@ application protocol some of which cannot be observed by the sender.
...
+STREAM or STREAM_DATA_BLOCKED frame if packets are lost or reordered.
 
-Before creating a stream, all lower-numbered streams of the same type MUST be
-created.  That means that receipt of a frame that would open a stream causes all
-lower-numbered streams of the same type to be opened in numeric order.  This
-ensures that the creation order for streams is consistent on both endpoints.
+Before creating a stream, all streams of the same type with lower-numbered
+stream IDs MUST be created.  This means that receipt of a frame that would open
+a stream causes all streams of the same type with lower-numbered stream IDs to
+be opened in numeric order.  This ensures that the creation order for streams is
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Drop the middle sentence of this paragraph.

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -640,19 +584,23 @@ STREAM_DATA_BLOCKED ({{frame-stream-data-blocked}}), and RST_STREAM
 
 A sender MUST NOT send any of these frames from a terminal state ("Data Recvd"
 or "Reset Recvd").  A sender MUST NOT send STREAM or STREAM_DATA_BLOCKED after
-sending a RST_STREAM; that is, in the "Reset Sent" state in addition to the
-terminal states.  A receiver could receive any of these frames in any state, but
-only due to the possibility of delayed delivery of packets carrying them.
+sending a RST_STREAM; that is, in the terminal states and in the "Reset Sent"
+state.
+
+A receiver could receive any of these three frames in any state, due to the
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: "these three frames" refers to something in another paragraph, so I think that your split should be reverted.

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -640,19 +584,23 @@ STREAM_DATA_BLOCKED ({{frame-stream-data-blocked}}), and RST_STREAM
...
 send STOP_SENDING in any state where it has not received a RST_STREAM frame;
 that is states other than "Reset Recvd" or "Reset Read".  However there is
-little value in sending a STOP_SENDING frame after all stream data has been
-received in the "Data Recvd" state.  A sender could receive these frames in any
-state as a result of delayed delivery of packets.
+little value in sending a STOP_SENDING frame in the "Data Recvd" state, since
+all stream data has been received.
+
+A sender could receive either of these two frames in any state as a result of
+delayed delivery of packets.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Same as previous.

File: draft-ietf-quic-transport.md [hunk 12]
```diff
@@ -725,102 +672,92 @@ data or a RST_STREAM frame has been received for the stream - that is, the
...
 # Flow Control {#flow-control}
 
-It is necessary to limit the amount of data that a sender may have outstanding
-at any time, so as to prevent a fast sender from overwhelming a slow receiver,
-or to prevent a malicious sender from consuming significant resources at a
-receiver.  To this end, QUIC employs a credit-based flow-control scheme similar
-to that in HTTP/2 {{?HTTP2}}.  A receiver advertises the number of bytes it is
-prepared to receive on a given stream and for the entire connection.  This leads
-to two levels of flow control in QUIC:
+It is necessary to limit the amount of data that a receiver may have to buffer,
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
It is necessary to limit the amount of data that a receiver could buffer,
```

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -725,102 +672,92 @@ data or a RST_STREAM frame has been received for the stream - that is, the
...
+so as to prevent a fast sender from overwhelming a slow receiver, or to prevent
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
to prevent a fast sender from overwhelming a slow receiver, or to prevent
```

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -725,102 +672,92 @@ data or a RST_STREAM frame has been received for the stream - that is, the
...
+with the Stream ID set appropriately.  A MAX_STREAM_DATA frame indicates the
+maximum absolute byte offset of a stream.  A receiver could use the current
+offset of data consumed to determine the flow control offset to be advertised.
+A receiver MAY send MAX_STREAM_DATA frames in multiple packets in order to make
+sure that the sender receives an update before running out of flow control
+credit, even if one of the packets is lost.
+
+A receiver advertises credit for a connection by sending a MAX_DATA frame, which
+indicates the maximum of the sum of the absolute byte offsets of all streams.  A
+receiver maintains a cumulative sum of bytes received on all contributing
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
receiver maintains a cumulative sum of bytes received on all
```

File: draft-ietf-quic-transport.md [hunk 15]
```diff
@@ -831,29 +768,46 @@ larger resource commitments at the receiver.  Thus there is a trade-off between
 resource commitment and overhead when determining how large a limit is
 advertised.
 
-A receiver MAY use an autotuning mechanism to tune the frequency and amount that
-it increases data limits based on a round-trip time estimate and the rate at
-which the receiving application consumes data, similar to common TCP
-implementations.
+A receiver MAY use an autotuning mechanism to tune the frequency and amount of
+advertised additional credit.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: The cut here might be too much.  I'd say either cut the whole paragraph, or keep it.  I think keeping it is best.

File: draft-ietf-quic-transport.md [hunk 16]
```diff
@@ -831,29 +768,46 @@ larger resource commitments at the receiver.  Thus there is a trade-off between
...
-If a sender runs out of flow control credit, it will be unable to send new
-data. That is, the sender is blocked. A blocked sender SHOULD send a
-STREAM_DATA_BLOCKED or DATA_BLOCKED frame.  A receiver uses these frames for
-debugging purposes.  A receiver MUST NOT wait for a STREAM_DATA_BLOCKED or
-DATA_BLOCKED frame before sending MAX_STREAM_DATA or MAX_DATA, since doing so
-will mean that a sender will be blocked for an entire round trip and the peer
-might never send a STREAM_DATA_BLOCKED or DATA_BLOCKED frame.
+If a sender runs out of flow control credit, it will be unable to send new data
+and is considered blocked.  It is generally considered best to not let the
+sender go into quiescence.  To avoid blocking a sender, and to reasonably
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
sender to become blocked.  To avoid blocking a sender, and to reasonably
```

File: draft-ietf-quic-transport.md [hunk 17]
```diff
@@ -831,29 +768,46 @@ larger resource commitments at the receiver.  Thus there is a trade-off between
...
 
-A sender sends a single STREAM_DATA_BLOCKED or DATA_BLOCKED frame only once when
-it reaches a data limit.  A sender SHOULD NOT send multiple STREAM_DATA_BLOCKED
-or DATA_BLOCKED frames for the same data limit, unless the original frame is
-determined to be lost.  Another STREAM_DATA_BLOCKED or DATA_BLOCKED frame can be
-sent after the data limit is increased.
+
+## Handling Stream Cancellation {#stream-cancellation}
+
+Given enough time, both endpoints must agree on flow control state, to avoid
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Endpoints need to eventually agree on the amount of flow control credit that has been consumed, to avoid either exceeding limits or deadlocking.

File: draft-ietf-quic-transport.md [hunk 18]
```diff
@@ -2762,6 +2719,28 @@ streams as necessary in outgoing packets without losing transmission efficiency
...
+Often, there will be limits on what can be transmitted as a result of connection
+flow control or the current congestion controller state.
+
+Giving preference to the transmission of its own management functions ensures
+that a protocol functions efficiently.  That is, prioritizing frames other than
+STREAM frames ensures that loss recovery, congestion control, and flow control
+operate effectively.
+
+CRYPTO frames SHOULD be prioritized over STREAM frames prior to the completion
+of the cryptographic handshake.  This includes the retransmission of the second
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: You already effectively said this in the last paragraph, so I think that you might want to drop this paragraph.

File: draft-ietf-quic-transport.md [hunk 19]
```diff
@@ -4838,9 +4816,11 @@ Stream Data:
 When a Stream Data field has a length of 0, the offset in the STREAM frame is
 the offset of the next byte that would be sent.
 
-The first byte in the stream has an offset of 0.  The largest offset delivered
-on a stream - the sum of the re-constructed offset and data length - MUST be
-less than 2^62.
+When new data is to be sent on a stream, a sender MUST set the encapsulating
+STREAM frame's offset field to the stream offset of the first byte of this new
+data.  The first byte in the stream has an offset of 0.  The largest offset
+delivered on a stream - the sum of the re-constructed offset and data length -
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]:
```suggestion
delivered on a stream - the sum of the offset and data length -
```

File: draft-ietf-quic-transport.md [hunk 20]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-There are two basic types of stream in QUIC.  Unidirectional streams carry data
-in one direction: from the initiator of the stream to its peer;
-bidirectional streams allow for data to be sent in both directions.  Different
-stream identifiers are used to distinguish between unidirectional and
-bidirectional streams, as well as to create a separation between streams that
-are initiated by the client and server (see {{stream-id}}).
+Streams in QUIC provide a lightweight, ordered byte-stream abstraction to an
+application. An alternative view of QUIC streams is as an elastic "message"
+abstraction, similar to the way ephemeral streams are used in SST
+{{?SST=DOI.10.1145/1282427.1282421}}, which may be a more appealing description
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: SST is a very dense description that may not be very helpful as an introduction for most users.

File: draft-ietf-quic-transport.md [hunk 21]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
+for some applications.
 
-Either type of stream can be created by either endpoint, can concurrently send
-data interleaved with other streams, and can be cancelled.
+QUIC does not provide any means of ensuring ordering between octets on different
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]:
```suggestion
QUIC does not provide any means of ensuring ordering between bytes on different
```

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]:
```suggestion
constraints (see {{flow-control}}) and a 62-bit stream offset limit.
```

File: draft-ietf-quic-transport.md (hunk 18 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: I think this paragraph is good to have since it is more specific. We can drop the previous paragraph, which is simply general advice.

File: draft-ietf-quic-transport.md [hunk 22]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-Endpoints MUST NOT exceed the limit set by their peer.  An endpoint that
-receives a STREAM frame with an ID greater than the limit it has sent MUST treat
-this as a stream error of type STREAM_LIMIT_ERROR ({{error-handling}}).
-
-A receiver cannot renege on an advertisement; that is, once a receiver
-advertises a stream limit using the MAX_STREAMS frame, advertising a smaller
-limit has no effect.  A receiver MUST ignore any MAX_STREAMS frame that does
-not increase the stream limit.
+Within each type, streams are expected to be created with numerically increasing
+Stream IDs.  A Stream ID that is used out of order results in all streams of
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]:
```suggestion
Stream IDs.  A stream opened an out of order Stream ID results in all streams of
```
(using a stream does not necessarily imply opening it, and therefore also not lower numbered streams)

File: draft-ietf-quic-transport.md [hunk 23]
```diff
@@ -240,178 +240,119 @@ x (*) ...
...
-flight of client handshake messages, that is, the TLS Finished and any client
-authentication messages.
-
-STREAM data in frames determined to be lost SHOULD be retransmitted before
-sending new data, unless application priorities indicate otherwise.
-Retransmitting lost stream data can fill in gaps, which allows the peer to
-consume already received data and free up the flow control window.
+resources to, the implementation SHOULD use the information provided by the
+application.  Failure to account for priority of streams can result in
+suboptimal application performance.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]:
```suggestion
suboptimal application performance, and potentially starvation or deadlocks.
```

File: draft-ietf-quic-transport.md [hunk 24]
```diff
@@ -420,8 +361,9 @@ conditions for opening a stream are slightly more complex for a bidirectional
 stream because the opening of either send or receive sides causes the stream
 to open in both directions.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: I'm not sure how to formulate this, but there is a difference between the endpoints and view, including the state machine, and the overall distributed view: A stream can only be opened by the initiator, but the first materialized view of a frame can be receiver side sending a packet to a bidi stream. This does not mean that the receiver is opening the stream, but to the receiver it looks like the stream is being opened (although it already knows that is is open from receiving a higher valued Stream ID).

In short, I think it is wrong to state that receive side can open a stream, but the correct formulation is not clear to me.

File: draft-ietf-quic-transport.md [hunk 25]
```diff
@@ -489,27 +431,30 @@ which can allow for better stream prioritization.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: Perhaps a note here on loose coupling between application stream handles and QUIC internal stream ID's?

**Comment by janaiyengar** [comment]
@mikkelfj : Sorry, missed your comments when I merged my PR. I'll take them in separately.

File: draft-ietf-quic-transport.md (hunk 24 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md]: Fair point, though as you note, it's not clear exactly what to replace this with. Feel free to suggest a PR for it if you come up with something here.
